### PR TITLE
KOGITO-3192: [DMN Designer] Multiple DRDs support - The undo/redo are lost when user changes between diagrams

### PR DESCRIPTION
**Pull request by danielzhe**
**JIRA**: [KOGITO-3192:](https://issues.redhat.com/browse/KOGITO-3192)

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

### CODE REVIEW

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by danielzhe** [comment]
You guys can check it before I improve the test coverage and check if the failed test is related. @karreiro @romartin

**Review by karreiro — COMMENTED** [review]
@danielzhe the approach here seems great to me! :-)

However, it's important to work on top of the https://github.com/kiegroup/kie-wb-common/pull/3498 PR. Otherwise, when it gets merged, maybe we have something broken here.

Also, I've cherry-picked [this commit](https://github.com/kiegroup/kie-wb-common/pull/3594/commits/c6cedfb071b2461aaafc1142543719831d8626e5) in the top of https://github.com/kiegroup/kie-wb-common/pull/3498, and I get this error:
Screenshot 2021-03-22 at 16 56 01

I don't think it's not something that will require a change of direction in the whole approach tho :-)

**Comment by danielzhe** [comment]
Hi! This PR was made in top of https://github.com/kiegroup/kie-wb-common/pull/3498. I just cherry-picked here and opened the PR. Did you rebuild all required repositories in https://github.com/kiegroup/kie-wb-common/pull/3498? @karreiro

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by karreiro** [comment]
@danielzhe is this PR ready for review? :-)

**Comment by jomarko** [comment]
Jenkins do fdb

**Comment by danielzhe** [comment]
Yes, @karreiro

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/command/ApplicationCommandManagerTest.java [hunk 1]
```diff
@@ -84,7 +83,7 @@ public void testInit() {
     @Test
     public void testStart() {
         tested.start();
-        verify(commandManager, times(1)).init(eq(session));
+        //verify(commandManager, times(1)).init(eq(session));
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/test/java/org/kie/workbench/common/stunner/core/client/command/ApplicationCommandManagerTest.java:86]: Please remove this commented line

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/drds/DMNDiagramsSession.java [hunk 2]
```diff
@@ -163,13 +177,59 @@ public DMNDiagramTuple getDiagramTuple(final String dmnDiagramElementId) {
         return getSessionState().getDMNDiagrams();
     }
 
+    public void onSessionDiagramOpenedEvent(final @Observes SessionDiagramOpenedEvent sessionDiagramOpenedEvent) {
+        loadHistoryForTheCurrentDiagram();
+    }
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/docks/navigator/drds/DMNDiagramsSession.java:182]: I didn't found unit tests of this method. Could you please point me to them?

**Review by jomarko — COMMENTED** [review]
### Manual Review
Hello @danielzhe. I am trying to review also VS Code extension using this PR. However there 'CTRL+Z' command doesn't work properly. In VS Code extension just one command can be un-done. For example add two nodes into the canvas. Press 'CTRL+Z' twice. Both nodes should disappear. however just one disappears.

It doesn't happen if I build 'kogito-tooling' master. 

Now it is question if this happens because:
- this PR changes
- changes this PR is based on, that are not present in 'kogito-tooling' master yet. Especially #3498

@ederign could you please advice if 'kogito-tooling' already contains changes introduced as #3498?

**Review by karreiro — COMMENTED** [review]
Thank you, @danielzhe!

I just had the issue that @jomarko has mentioned. Here's the `.vsix` plugin (it was built based on this PR rebased with master): [plugin](https://www.dropbox.com/s/rzqol327dh1b8hb/danielzhe-KOGITO-3192.vsix?dl=0).

**Comment by romartin** [comment]
hey guys,

It also looks good for me. I'm now investigating the fix for reopening DRDs, and I want to be sure that I will not brake this, once merged, so go ahead with it, but depending on which PR gets in first, maybe we'l lhave to sync @danielzhe . 
Anyway no worries, will let you know my progress and try this fix also on top of my PR, to ensure we'll get rid of both issues.

So @karreiro @jomarko once you're feeling happy with this PR, go ahead and merge, I'll keep an eye here and ensure it'll work also on my fix work.

Thanks guys

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by danielzhe** [comment]
Changes applied. Ready for your review, @jomarko and @karreiro

**Comment by jomarko** [comment]
@danielzhe 

there are test failures and seems be related to our changes
org.kie.workbench.common.dmn.client.session.DMNEditorSessionTest.testGetCommandRegistry
org.kie.workbench.common.stunner.core.client.session.command.impl.UndoSessionCommandTest.testExecuteWithErrors
org.kie.workbench.common.stunner.core.client.session.command.impl.UndoSessionCommandTest.testExecuteSuccess
org.kie.workbench.common.stunner.core.client.session.command.impl.UndoSessionCommandTest.checkRespondsToExpectedKeys
org.kie.workbench.common.stunner.kogito.client.session.command.impl.KogitoUndoSessionCommandTest.testBindCommandOutsideKogito

**Comment by jomarko** [comment]
@danielzhe the issue I previously found in VS code (https://github.com/kiegroup/kie-wb-common/pull/3594#pullrequestreview-625194663) can not be reproduced anymore. Also business central seems working, please fix tests and this can be merged. Thank you.

**Review by karreiro — COMMENTED** [review]
Thank you, @danielzhe! LGTM, and works fine as well.

I'll just wait for the tests before leaving the final approval :-)

**Review by jomarko — APPROVED** [review]

**Review by karreiro — APPROVED** [review]
Thank you, @danielzhe!

**Review by romartin — APPROVED** [review]

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please fdb.

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by karreiro** [comment]
Jenkins do fdb.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by danielzhe** [comment]
Jenkins do fdb

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by danielzhe** [comment]
Jenkins please retest this

**Comment by danielzhe** [comment]
jenkins do fdb

**Comment by karreiro** [comment]
jenkins do fdb

**Comment by karreiro** [comment]
jenkins do fdb

**Pull request merged**